adblock2privoxy compilation fail with GHC 8.4.3

Adblock2privoxy fails to compile  with GHC 8.4.3. Please take a look would fixing it be possible.
```
>>> Configuring source in /tmp/portage/net-proxy/adblock2privoxy-9999-r2/work/adblock2privoxy-9999/adblock2privoxy ...
 * Using cabal-2.2.0.1.
 * Prepending /usr/lib64/ghc-8.4.3 to LD_LIBRARY_PATH
/usr/bin/ghc -package Cabal-2.2.0.1 --make /tmp/portage/net-proxy/adblock2privoxy-9999-r2/work/adblock2privoxy-9999/adblock2privoxy/Setup.hs -j5 +RTS -A256M -qb0 -RTS -threaded -dynamic -o setup
[1 of 1] Compiling Main             ( /tmp/portage/net-proxy/adblock2privoxy-9999-r2/work/adblock2privoxy-9999/adblock2privoxy/Setup.hs, /tmp/portage/net-proxy/adblock2privoxy-9999-r2/work/adblock2privoxy-9999/adblock2privoxy/Setup.o )
Linking setup ...
./setup configure --ghc --prefix=/usr --with-compiler=/usr/bin/ghc --with-hc-pkg=/usr/bin/ghc-pkg --prefix=/usr --libdir=/usr/lib64 --libsubdir=adblock2privoxy-9999/ghc-8.4.3 --datadir=/usr/share/ --datasubdir=adblock2privoxy-9999/ghc-8.4.3 --ghc-options=-j5 +RTS -A256M -qb0 -RTS --with-ar=x86_64-pc-linux-gnu-ar --ghc-option=-optc-march=native --ghc-option=-optc-O2 --ghc-option=-optc-pipe --ghc-option=-optc-fgcse-sm --ghc-option=-optc-fgcse-las --ghc-option=-optc-fgcse-after-reload --ghc-option=-optc-ftree-vectorize --ghc-option=-optl-Wl,-O1 --ghc-option=-optl-Wl,--hash-style=both --ghc-option=-optl-Wl,--as-needed --ghc-option=-optl-Wl,--relax --ghc-option=-optl-Wl,--sort-common --disable-executable-stripping --docdir=/usr/share/doc/adblock2privoxy-9999-r2 --verbose --enable-shared --enable-executable-dynamic --sysconfdir=/etc --disable-library-stripping
Using Parsec parser
Configuring adblock2privoxy-1.4.2...
Dependency MissingH -any: using MissingH-1.4.0.1
Dependency base ==4.*: using base-4.11.1.0
Dependency case-insensitive -any: using case-insensitive-1.2.0.11
Dependency containers -any: using containers-0.5.11.0
Dependency directory -any: using directory-1.3.1.5
Dependency filepath -any: using filepath-1.4.2
Dependency http-conduit -any: using http-conduit-2.3.1
Dependency mtl -any: using mtl-2.2.2
Dependency network >=2.4: using network-2.6.3.2
Dependency network-uri -any: using network-uri-2.6.1.0
Dependency old-locale >=1.0: using old-locale-1.0.0.7
Dependency parsec -any: using parsec-3.1.13.0
Dependency parsec-permutation >=0.1.2.0: using parsec-permutation-0.1.2.0
Dependency strict >=0.3: using strict-0.3.2
Dependency text >=0.11: using text-1.2.3.0
Dependency time >=1.4: using time-1.8.0.2
Source component graph: component exe:adblock2privoxy
Configured component graph:
    component adblock2privoxy-1.4.2-2X1oyazVohi4OGBtlFZzjK-adblock2privoxy
        include base-4.11.1.0
        include parsec-3.1.13.0
        include mtl-2.2.2
        include containers-0.5.11.0
        include filepath-1.4.2
        include directory-1.3.1.5
        include MissingH-1.4.0.1-GvxKFm8JIQLIx04ZzIEPgN
        include parsec-permutation-0.1.2.0-3zDFfngbNb4CTcNShrLuUV
        include time-1.8.0.2
        include old-locale-1.0.0.7-26K7wLFR2jK44UeOklvTCh
        include strict-0.3.2-FUXfVuv1I9pDkeq770AklR
        include network-2.6.3.2-6UULJJw3egYFPRHsaFrlj6
        include http-conduit-2.3.1-InHJXoxZeVnHqoo92GzyfZ
        include text-1.2.3.0
        include network-uri-2.6.1.0-AstEwZoXrlUJQq4VkxaVo9
        include case-insensitive-1.2.0.11-3kxnS3DflzuLJ3paEvK126
Linked component graph:
    unit adblock2privoxy-1.4.2-2X1oyazVohi4OGBtlFZzjK-adblock2privoxy
        include base-4.11.1.0
        include parsec-3.1.13.0
        include mtl-2.2.2
        include containers-0.5.11.0
        include filepath-1.4.2
        include directory-1.3.1.5
        include MissingH-1.4.0.1-GvxKFm8JIQLIx04ZzIEPgN
        include parsec-permutation-0.1.2.0-3zDFfngbNb4CTcNShrLuUV
        include time-1.8.0.2
        include old-locale-1.0.0.7-26K7wLFR2jK44UeOklvTCh
        include strict-0.3.2-FUXfVuv1I9pDkeq770AklR
        include network-2.6.3.2-6UULJJw3egYFPRHsaFrlj6
        include http-conduit-2.3.1-InHJXoxZeVnHqoo92GzyfZ
        include text-1.2.3.0
        include network-uri-2.6.1.0-AstEwZoXrlUJQq4VkxaVo9
        include case-insensitive-1.2.0.11-3kxnS3DflzuLJ3paEvK126
Ready component graph:
    definite adblock2privoxy-1.4.2-2X1oyazVohi4OGBtlFZzjK-adblock2privoxy
        depends base-4.11.1.0
        depends parsec-3.1.13.0
        depends mtl-2.2.2
        depends containers-0.5.11.0
        depends filepath-1.4.2
        depends directory-1.3.1.5
        depends MissingH-1.4.0.1-GvxKFm8JIQLIx04ZzIEPgN
        depends parsec-permutation-0.1.2.0-3zDFfngbNb4CTcNShrLuUV
        depends time-1.8.0.2
        depends old-locale-1.0.0.7-26K7wLFR2jK44UeOklvTCh
        depends strict-0.3.2-FUXfVuv1I9pDkeq770AklR
        depends network-2.6.3.2-6UULJJw3egYFPRHsaFrlj6
        depends http-conduit-2.3.1-InHJXoxZeVnHqoo92GzyfZ
        depends text-1.2.3.0
        depends network-uri-2.6.1.0-AstEwZoXrlUJQq4VkxaVo9
        depends case-insensitive-1.2.0.11-3kxnS3DflzuLJ3paEvK126
Using Cabal-2.2.0.1 compiled by ghc-8.4
Using compiler: ghc-8.4.3
Using install prefix: /usr
Executables installed in: /usr/bin
Libraries installed in: /usr/lib64/adblock2privoxy-9999/ghc-8.4.3
Dynamic Libraries installed in: /usr/lib64/x86_64-linux-ghc-8.4.3
Private executables installed in:
/usr/libexec/x86_64-linux-ghc-8.4.3/adblock2privoxy-1.4.2
Data files installed in: /usr/share/adblock2privoxy-9999/ghc-8.4.3
Documentation installed in: /usr/share/doc/adblock2privoxy-9999-r2
Configuration files installed in: /etc
Using alex version 3.2.4 found on system at: /usr/bin/alex
Using ar given by user at: /usr/bin/x86_64-pc-linux-gnu-ar
Using c2hs version 0.28.5 found on system at: /usr/bin/c2hs
Using cpphs version 1.20.8 found on system at: /usr/bin/cpphs
No doctest found
Using gcc version 8.2.0 found on system at: /usr/bin/x86_64-pc-linux-gnu-gcc
Using ghc version 8.4.3 given by user at: /usr/bin/ghc
Using ghc-pkg version 8.4.3 given by user at: /usr/bin/ghc-pkg
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.20.0 found on system at: /usr/bin/haddock-ghc-8.4.3
Using happy version 1.19.9 found on system at: /usr/bin/happy
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.67 found on system at: /usr/bin/hpc
Using hsc2hs version 0.68.3 found on system at: /usr/bin/hsc2hs
No hscolour found
No jhc found
Using ld found on system at: /usr/bin/x86_64-pc-linux-gnu-ld
No lhc found
No lhc-pkg found
Using pkg-config version 0.29.2 found on system at: /usr/bin/pkg-config
Using runghc version 8.4.3 found on system at: /usr/bin/runghc
Using strip version 2.30 found on system at: /usr/bin/strip
Using tar found on system at: /bin/tar
No uhc found
>>> Source configured.
>>> Compiling source in /tmp/portage/net-proxy/adblock2privoxy-9999-r2/work/adblock2privoxy-9999/adblock2privoxy ...
./setup build
Preprocessing executable 'adblock2privoxy' for adblock2privoxy-1.4.2..
Building executable 'adblock2privoxy' for adblock2privoxy-1.4.2..
[ 1 of 18] Compiling InputParser      ( src/InputParser.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/InputParser.dyn_o )
[ 2 of 18] Compiling Network          ( src/Network.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Network.dyn_o )
[ 3 of 18] Compiling Paths_adblock2privoxy ( dist/build/adblock2privoxy/autogen/Paths_adblock2privoxy.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Paths_adblock2privoxy.dyn_o )
[ 4 of 18] Compiling PopupBlocker     ( src/PopupBlocker.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/PopupBlocker.dyn_o )
[ 5 of 18] Compiling ProgramOptions   ( src/ProgramOptions.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/ProgramOptions.dyn_o )
[ 6 of 18] Compiling SourceInfo       ( src/SourceInfo.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/SourceInfo.dyn_o )
[ 7 of 18] Compiling Statistics       ( src/Statistics.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Statistics.dyn_o )
[ 8 of 18] Compiling Task             ( src/Task.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Task.dyn_o )
[ 9 of 18] Compiling UrlBlocker[boot] ( src/UrlBlocker.hs-boot, dist/build/adblock2privoxy/adblock2privoxy-tmp/UrlBlocker.dyn_o-boot )
[10 of 18] Compiling Templates        ( src/Templates.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Templates.dyn_o )
[11 of 18] Compiling OptionsConverter ( src/OptionsConverter.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/OptionsConverter.dyn_o )
[12 of 18] Compiling Utils            ( src/Utils.hs, dist/build/adblock2privoxy/adblock2privoxy-tmp/Utils.dyn_o )

src/Utils.hs:60:10: error:
    • Could not deduce (Semigroup (ZipListM a))
        arising from the superclasses of an instance declaration
      from the context: Monoid a
        bound by the instance declaration at src/Utils.hs:60:10-40
    • In the instance declaration for ‘Monoid (ZipListM a)’
   |
60 | instance Monoid a => Monoid (ZipListM a) where
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 * ERROR: net-proxy/adblock2privoxy-9999-r2::bigvo failed (compile phase):
 *   setup build failed
 * 
 * Call stack:
 *     ebuild.sh, line  124:  Called src_compile
 *   environment, line 3350:  Called haskell-cabal_src_compile
 *   environment, line 2537:  Called cabal_src_compile
 *   environment, line  911:  Called cabal-build
 *   environment, line  675:  Called die
 * The specific snippet of code:
 *       ./setup "$@" || die "setup build failed"
 * 

>>> Failed to emerge net-proxy/adblock2privoxy-9999-r2, Log file:
```

Per #26, I'm going to try to address this now. I'll need your help.

Rather than go to the trouble of creating a new branch, I'll use the existing one and show you the diffs, unless this becomes too burdensome.

Here's what I've done:

1. Download and install latest `ghc-8.4.3` from https://www.haskell.org/platform/
2. rsync the latest branch to `/tmp/adblock2privoxy`

```
mkdir /tmp/adblock2privoxy
rsync essandess/adblock2privoxy/adblock2privoxy* /tmp/adblock2privoxy
cd /tmp/adblock2privoxy
vi stack.yaml
```

3. Make these changes:

```
resolver: ghc-8.4.3
system-ghc: true
```

4. Run the stack `setup` and `install` commands, making sure that it find the right binaries:

```
bash -c 'export PATH=/usr/bin:/usr/local/bin:$PATH ; export STACK_ROOT=/tmp/adblock2privoxy/.stack ; ( cd /tmp/adblock2privoxy/adblock2privoxy ; stack setup --allow-different-user ; stack install --local-bin-path /tmp/adblock2privoxy --allow-different-user )'
```

First error look like below. Keep adding stuff to `extra-deps` in `stack.yaml` as suggested, including `allow-newer: true`.
```
Error: While constructing the build plan, the following exceptions were
encountered:

In the dependencies for adblock2privoxy-1.4.2:
    MissingH needed, but the stack configuration has no specified version 
             (latest matching version is 1.4.0.1)
    http-conduit needed, but the stack configuration has no specified version 
                 (latest matching version is 2.3.2)
    old-locale must match >=1.0, but the stack configuration has no specified
               version  (latest matching version is 1.0.0.7)
    strict must match >=0.3, but the stack configuration has no specified
           version  (latest matching version is 0.3.2)
needed since adblock2privoxy is a build target.

Some different approaches to resolving this:

  * Consider trying 'stack solver', which uses the cabal-install solver to
    attempt to find some working build configuration. This can be convenient
    when dealing with many complicated constraint errors, but results may be
    unpredictable.

  * Recommended action: try adding the following to your extra-deps
    in /private/tmp/adblock2privoxy/adblock2privoxy/stack.yaml:

- MissingH-1.4.0.1
- http-conduit-2.3.2
- old-locale-1.0.0.7
- strict-0.3.2
```

5. Try again with the following `stack.yaml` (`diff`). It starts to build.
```
4c4
< resolver: lts-8.23
---
> resolver: ghc-8.4.3
9a10,11
> allow-newer: true
> 
11c13,64
< extra-deps: [parsec-permutation-0.1.2.0]
---
> extra-deps:
> - parsec-permutation-0.1.2.0
> - MissingH-1.4.0.1
> - http-conduit-2.3.2
> - old-locale-1.0.0.7
> - strict-0.3.2
> - aeson-1.4.0.0
> - conduit-1.3.0.3
> - conduit-extra-1.3.0
> - hslogger-1.2.10
> - http-client-0.5.13.1
> - http-client-tls-0.3.5.3
> - http-types-0.12.1
> - old-time-1.1.0.3
> - resourcet-1.2.1
> - unliftio-core-0.1.2.0
> - base-compat-0.10.4
> - blaze-builder-0.4.1.0
> - connection-0.2.8
> - cookie-0.4.4
> - cryptonite-0.25
> - data-default-class-0.1.2.0
> - dlist-0.8.0.5
> - exceptions-0.10.0
> - memory-0.14.16
> - mime-types-0.1.0.8
> - mono-traversable-1.0.9.0
> - streaming-commons-0.2.1.0
> - tagged-0.8.6
> - th-abstraction-0.2.8.0
> - time-locale-compat-0.1.1.5
> - tls-1.4.1
> - typed-process-0.2.3.0
> - uuid-types-1.0.3
> - asn1-encoding-0.9.5
> - asn1-types-0.3.2
> - basement-0.0.8
> - byteable-0.1.1
> - cereal-0.5.7.0
> - foundation-0.0.21
> - socks-0.5.6
> - time-1.4.2
> - transformers-compat-0.6.2
> - vector-algorithms-0.7.0.4
> - x509-1.7.4
> - x509-store-1.6.6
> - x509-system-1.6.6
> - x509-validation-1.6.10
> - asn1-parse-0.9.4
> - deepseq-1.3.0.2
> - hourglass-0.2.12
> - pem-0.2.4
20c73
< # system-ghc: true
---
> system-ghc: true
```

6. Finally, encounter this compile-time error:
```
stack will use the GHC on your PATH
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
jWARNING: Ignoring out of range dependency (allow-newer enabled): base-4.11.1.0. deepseq requires: >=4.3 && <4.8
WARNING: Ignoring out of range dependency (allow-newer enabled): transformers-0.5.5.0. transformers-compat requires: >=0.3 && ==0.2.*
WARNING: Ignoring out of range dependency (allow-newer enabled): time-1.4.2. cookie requires: >=1.5
jj[1 of 2] Compiling Main             ( /private/tmp/adblock2privoxy/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /private/tmp/adblock2privoxy/.stack/setup-exe-src/setup-mPHDZzAJ.o )
…

parsec-3.1.13.0: copy/register                                         
cryptonite-0.25: copy/register                  
Progress 49/75                 

--  While building custom Setup.hs for package cookie-0.4.4 using:
      /private/tmp/adblock2privoxy/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_2.2.0.1_ghc-8.4.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-2.2.0.1 build --ghc-options " -ddump-hi -ddump-to-file -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
    Logs have been written to: /private/tmp/adblock2privoxy/adblock2privoxy/.stack-work/logs/cookie-0.4.4.log

    Configuring cookie-0.4.4...
    clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
    Preprocessing library for cookie-0.4.4..
    Building library for cookie-0.4.4..
    [1 of 1] Compiling Web.Cookie       ( Web/Cookie.hs, .stack-work/dist/x86_64-osx/Cabal-2.2.0.1/build/Web/Cookie.o )
    
    /private/var/folders/w3/pqgw02vs1zjd_nd0gdc1ctf80000gp/T/stack1259/cookie-0.4.4/Web/Cookie.hs:45:78: error:
        Module ‘Data.Time’ does not export ‘parseTimeM’
       |
    45 | import Data.Time (UTCTime (UTCTime), toGregorian, fromGregorian, formatTime, parseTimeM, defaultTimeLocale)
       |                                                                              ^^^^^^^^^^
    
    /private/var/folders/w3/pqgw02vs1zjd_nd0gdc1ctf80000gp/T/stack1259/cookie-0.4.4/Web/Cookie.hs:45:90: error:
        Module ‘Data.Time’ does not export ‘defaultTimeLocale’
       |
    45 | import Data.Time (UTCTime (UTCTime), toGregorian, fromGregorian, formatTime, parseTimeM, defaultTimeLocale)
       |                                                                                          ^^^^^^^^^^^^^^^^^
    
```

A few things are clearly wrong here. Having to add all those `extra-dep`'s and `allow-newer: true` looks wrong. The compile error at Module `Data.Time` is different from your compile error.

I'm not sure which issue to tackle. Any insight based on your experience?

--- a/adblock2privoxy/src/Utils.hs
+++ b/adblock2privoxy/src/Utils.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 module Utils (
 Struct2 (..),
 Struct3 (..),
@@ -57,9 +58,18 @@ getZipListM = getZipList.getZipList'
 zipListM :: [a] -> ZipListM a
 zipListM = ZipListM . ZipList
 
+#if (MIN_VERSION_base(4,11,0))
+instance Semigroup a => Semigroup (ZipListM a) where
+  x <> y = (<>) <$> x <*> y
+#endif
+
 instance Monoid a => Monoid (ZipListM a) where
   mempty = pure mempty
+#if (MIN_VERSION_base(4,11,0))
+  mappend = (<>)
+#else
   mappend x y = mappend <$> x <*> y
+#endif
 
 class Struct2 f where
         struct2 :: a1 -> a2 -> f a1 a2

--- a/adblock2privoxy/stack.yaml
+++ b/adblock2privoxy/stack.yaml
@@ -1,7 +1,7 @@
 # For more information, see: https://github.com/commercialhaskell/stack/blob/release/doc/yaml_configuration.md
 
 # Specifies the GHC version and set of packages available (e.g., lts-3.5, nightly-2015-09-21, ghc-7.10.2)
-resolver: lts-8.23
+resolver: lts-12.9
 
 # Local packages, usually specified by relative directory name
 packages:

